orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/jedis/AbstractRedisExecutionRepository.java

```diff
@@ -0,0 +1,343 @@
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.jackson.OrcaObjectMapper;
+import com.netflix.spinnaker.orca.pipeline.model.Execution;
+import com.netflix.spinnaker.orca.pipeline.model.Execution.PausedDetails;
+import com.netflix.spinnaker.orca.pipeline.model.Execution.ExecutionType;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import com.netflix.spinnaker.orca.pipeline.model.Task;
+import com.netflix.spinnaker.orca.pipeline.persistence.ExecutionNotFoundException;
+import com.netflix.spinnaker.orca.pipeline.persistence.ExecutionRepository;
+import groovy.util.logging.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import redis.clients.jedis.JedisCommands;
+import rx.Scheduler;
+
+import javax.annotation.Nonnull;
+
+import static com.netflix.spinnaker.orca.pipeline.model.Execution.ExecutionType.PIPELINE;
+import static java.lang.System.currentTimeMillis;
+import static java.lang.String.format;
+
+@Slf4j
+abstract class AbstractRedisExecutionRepository implements ExecutionRepository {
+
+  static final TypeReference<List<Task>> LIST_OF_TASKS =
+    new TypeReference<List<Task>>() {
+    };
+  static final TypeReference<Map<String, Object>> MAP_STRING_TO_OBJECT =
+    new TypeReference<Map<String, Object>>() {
+    };
+  final RedisClientDelegate redisClientDelegate;
+  final Optional<RedisClientDelegate> previousRedisClientDelegate;
+  protected final ObjectMapper mapper = OrcaObjectMapper.newInstance();
+  final int chunkSize;
+  final Scheduler queryAllScheduler;
+  final Scheduler queryByAppScheduler;
+  private final Registry registry;
+
+  AbstractRedisExecutionRepository(
+    Registry registry,
+    RedisClientDelegate redisClientDelegate,
+    Optional<RedisClientDelegate> previousRedisClientDelegate,
+    Scheduler queryAllScheduler,
+    Scheduler queryByAppScheduler,
+    Integer threadPoolChunkSize
+  ) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.previousRedisClientDelegate = previousRedisClientDelegate;
+    this.queryAllScheduler = queryAllScheduler;
+    this.queryByAppScheduler = queryByAppScheduler;
+    this.chunkSize = threadPoolChunkSize;
+    this.registry = registry;
+  }
+
+  @Override
+  public void cancel(@Nonnull String id) {
+    cancel(id, null, null);
+  }
+
+  @Override
+  public void cancel(@Nonnull String id, String user, String reason) {
+    String key = fetchKey(id);
+
+    getRedisDelegateForId(key).withCommandsClient(c -> {
+      HashMap<String, String> data = new HashMap<>();
+      data.put("canceled", "true");
+      if (StringUtils.isNotEmpty(user)) {
+        data.put("canceledBy", user);
+      }
+      if (StringUtils.isNotEmpty(reason)) {
+        data.put("cancellationReason", reason);
+      }
+      ExecutionStatus currentStatus = ExecutionStatus.valueOf(c.hget(key, "status"));
+      if (currentStatus == ExecutionStatus.NOT_STARTED) {
+        data.put("status", ExecutionStatus.CANCELED.name());
+      }
```

Comment:
@robfletcher I'd love to get a summary of why we don't always set the status to `CANCELED` here.

Reply:
Status is set to `CANCELLED` (so long as `executionStatus != TERMINAL`) via an `afterExecution` listener in `ExecutionPropagationListener`. I think this makes sense, a task could be executing on an orca instance other than the one handling the cancel request. That task will still complete, so seems correct to set pipeline status to `CANCELED` at that time.

Reply:
Yeah, the idea is you don't want to interrupt a graceful stop of the pipeline.